README.md
# Overwolf - Get supported features

This module lets you fetch the supported features that you can register for each game (that supports it).

This works best from within a Overwolf application, if you want to register actual features/events from games you play, it can't fetch events from i.e. the web.

## How to use (in Overwolf applications)

Load the library either by importing it or loading it before the code that needs to use it.

```js
/* Don't register listeners like this.. */
overwolf.games.onGameLaunched.addListener((info) => {
  let supportedFeatures = overwolf.games.getSupportedFeatures(info.classId);
  console.log(supportedFeatures); // { game: "APEX", events: [ ... all the events ] }
});

/* You can also fetch the supported features like this */
let supportedFeatures = overwolf.games.getSupportedFeatures(
  OverwolfGameWithEventSupport.Overwatch
);

/* And if the plugin should be outdated on what features are available, you can always use this */

overwolf.games.getSupportedFeaturesOnline(
  OverwolfGameWithEventSupport.APEX,
  (features) => {
    console.log(features);
  }
);
```

## How to use (in javascript outside of Overwolf)

```js
let {
  getSupportedFeatures,
  getSupportedFeaturesOnline,
  OverwolfGameWithEventSupport,
} = require("overwolf-games-features");

let supportedFeatures = getSupportedFeatures(
  OverwolfGameWithEventSupport.Overwatch
);

console.log(supportedFeatures);
```

## For more information about Overwolf and their applications

**Developers** - Join their [Official Developer Discord](https://discord.gg/overwolf-developers)

**End users** - Join their [Official Discord](https://discord.gg/overwolf)

**Anyone with questions about this module**

Feel free to join [my Discord](https://discord.gg/6fP8vWW)


CHANGELOG.md
# Changelog

## 1.8.2

- feat: Added more supported features for CoD: Warzone (assist, roster)

## 1.8.1

- fix: Added more supported features for FFXIV

## 1.8.0

- feat: Added support for Call of Duty: Warzone, Final Fantasy XIV and Warframe

## 1.7.0

- feat: Added support for Minecraft

## 1.6.0

- feat: Added the method `getSupportedFeaturesOnline`, that fetches the available features from Overwolf-servers (uses callback)

## 1.5.0

- feat(Fortnite): Added `killer`-feature

## 1.4.0

- feat(PUBG Lite): Added `team`-feature
- fix(Valorant): Fixed supported features

## 1.3.0

- feat: Added support for [Overwatch](https://overwolf.github.io/docs/api/overwolf-games-events-overwatch)

## 1.2.0

- feat: Added the new events (possibly features) for Valorant

## 1.1.1

- fix: Changed the output type when calling the `getSupportedFeatures`-method.

## 1.1.0

- fix: Editing output to contain both game and events
- fix: Changed how we set the method to the overwolf.games-namespace

## 1.0.2

Dunno

## 1.0.0

- feat: First release, only returned the array of supported features


test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("./index");
let features = index_1.getSupportedFeatures(index_1.OverwolfGameWithEventSupport.APEX);
console.log(features);
index_1.getSupportedFeaturesOnline(index_1.OverwolfGameWithEventSupport.APEX, data => {
    console.log(data);
});


src/test.ts
import { OverwolfGameWithEventSupport, getSupportedFeatures, getSupportedFeaturesOnline } from "./index";

let features = getSupportedFeatures(OverwolfGameWithEventSupport.APEX);
console.log(features);

getSupportedFeaturesOnline(OverwolfGameWithEventSupport.APEX, data => {
  console.log(data);
});


src/index.ts
'use strict';

/**
 * An enum that contains all games that support events through Overwolf
 */
export const OverwolfGameWithEventSupport = {
  APEX: 21566,
  CSGO: 7764,
  DOTA2: 7314,
  DOTAUnderlords: 21586,
  EscapeFromTarkov: 21634,
  Fortnite: 21216,
  Hearthstone: 9898,
  HeroesOfTheStorm: 10624,
  LeagueOfLegends: 5426,
  LegendsOfRuneterra: 21620,
  MTGArena: 21308,
  Overwatch: 10844,
  PUBG: 10906,
  PUBGLite: 21598,
  PathOfExile: 7212,
  RainbowSixSiege: 10826,
  RocketLeague: 10798,
  Splitgate: 21404,
  StarCraft2: 5855,
  TeamfightTactics: 5426,
  Valorant: 21640,
  WorldOfTanks: 6365,
  WorldOfWarcraft: 765,
  WorldOfWarships: 10746
};

/**
 * Enum that contains the supported events per game, these will automatically register with setRequiredFeatures when the game is detected
 */
export const OverwolfGameSupportedEvents = {
  APEX: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'me',
    'team',
    'kill',
    'damage',
    'death',
    'revive',
    'match_state',
    'match_info',
    'inventory',
    'location',
    'match_summary',
    'roster',
    'rank',
    'kill_feed'
  ],
  CSGO: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'match_info',
    'kill',
    'death',
    'assist',
    'headshot',
    'round_start',
    'match_start',
    'match_info',
    'match_end',
    'team_round_win',
    'bomb_planted',
    'bomb_change',
    'reloading',
    'fired',
    'weapon_change',
    'weapon_acquired',
    'info',
    'roster',
    'player_activity_change',
    'team_set',
    'replay',
    'counters',
    'mvp',
    'scoreboard',
    'kill_feed'
  ],
  DOTA2: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'game_state_changed',
    'match_state_changed',
    'match_detected',
    'daytime_changed',
    'clock_time_changed',
    'ward_purchase_cooldown_changed',
    'match_ended',
    'kill',
    'assist',
    'death',
    'cs',
    'xpm',
    'gpm',
    'gold',
    'hero_leveled_up',
    'hero_respawned',
    'hero_buyback_info_changed',
    'hero_boughtback',
    'hero_health_mana_info',
    'hero_status_effect_changed',
    'hero_attributes_skilled',
    'hero_ability_skilled',
    'hero_ability_used',
    'hero_ability_cooldown_changed',
    'hero_ability_changed',
    'hero_item_cooldown_changed',
    'hero_item_changed',
    'hero_item_used',
    'hero_item_consumed',
    'hero_item_charged',
    'match_info',
    'roster',
    'party',
    'error',
    'hero_pool',
    'me',
    'game'
  ],
  DOTAUnderlords: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'match_info'
  ],
  EscapeFromTarkov: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'match_info',
    'game_info'
  ],
  Fortnite: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'kill',
    'killed',
    'killer',
    'revived',
    'death',
    'match',
    'match_info',
    'rank',
    'me',
    'phase',
    'location',
    'team',
    'items',
    'counters'
  ],
  Hearthstone: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'scene_state',
    'collection',
    'decks',
    'match',
    'match_info'
  ],
  HeroesOfTheStorm: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'match_info',
    'me',
    'game_info',
    'roster',
    'death',
    'kill'
  ],
  LeagueOfLegends: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'live_client_data',
    'matchState',
    'match_info',
    'death',
    'respawn',
    'abilities',
    'kill',
    'assist',
    'gold',
    'minions',
    'summoner_info',
    'gameMode',
    'teams',
    'level',
    'announcer',
    'counters',
    'damage',
    'heal'
  ],
  LegendsOfRuneterra: [
    'game_client_data'
  ],
  MTGArena: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'game_info',
    'match_info'
  ],
  Overwatch: [
    'gep_internal',
    'game_info',
    'match_info',
    'kill',
    'death'
  ],
  PUBG: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'kill',
    'revived',
    'death',
    'killer',
    'match',
    'rank',
    'counters',
    'location',
    'me',
    'team',
    'phase',
    'map',
    'roster'
  ],
  PUBGLite: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'kill',
    'revived',
    'death',
    'killer',
    'match',
    'rank',
    'me',
    'phase',
    'map',
    'team_feed',
    'team'
  ],
  PathOfExile: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'kill',
    'death',
    'me',
    'match_info'
  ],
  RainbowSixSiege: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'game_info',
    'match',
    'match_info',
    'roster',
    'kill',
    'death',
    'me'
  ],
  RocketLeague: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'stats',
    'teamGoal',
    'opposingTeamGoal',
    'match',
    'roster',
    'me',
    'match_info'
  ],
  Splitgate: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'game_info',
    'match_info',
    'player',
    'location',
    'match',
    'feed',
    'connection',
    'kill',
    'death',
    'portal',
    'assist'
  ],
  StarCraft2: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'match_info'
  ],
  TeamfightTactics: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'live_client_data',
    'me',
    'match_info',
    'roster',
    'store',
    'board',
    'bench',
    'carousel'
  ],
  Valorant: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'me',
    'game_info',
    'match_info',
    'kill',
    'death'
  ],
  WorldOfTanks: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'kill',
    'death',
    'game_info',
    'match_info'
  ],
  WorldOfWarcraft: [
    'game_info'
  ],
  WorldOfWarships: [
    /**
     * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
     */
    'gep_internal',
    'game_info',
    'account_info',
    'match',
    'match_info',
    'kill',
    'death'
  ]
};

/**
 * @typedef {Object} OverwolfGameWithEvents
 * @property {string} game
 * @property {string[]} events
 */


/**
 * Gets the supported events for the specified classId
 * @param {Number} classId This is the classId you get from the gameInfo-object
 * @returns {OverwolfGameWithEvents} Returns an array of supported events if there's a game that corresponds to the classId, otherwise null
 */
export function getSupportedFeatures(classId) {
  let game = Object.keys(OverwolfGameWithEventSupport).find(key => OverwolfGameWithEventSupport[key] == classId);
  if (game) {
    return {
      game: game,
      events: OverwolfGameSupportedEvents[game]
    };
  }
  return null;
}

/**
 * Gets the supported events for the specified classId
 * @param {Number} classId This is the classId you get from the gameInfo-object
 * @param {Function|null} callback This is what you use to manage the data returned
 */
export function getSupportedFeaturesOnline(classId, callback?) {
  let game = Object.keys(OverwolfGameWithEventSupport).find(key => OverwolfGameWithEventSupport[key] == classId);
  if (game) {
    fetch(`https://game-events-status.overwolf.com/${classId}_prod.json`)
      .then(resp => {
        return resp.json()
      })
      .then(json => {
        if (!!callback) {
          callback({
            game: game,
            events: json.features.map(f => f.name)
          })
        }
      }
      )
      .catch(err => console.log(err));
  }
}

if (typeof overwolf !== 'undefined' && typeof overwolf.games !== 'undefined') {
  (overwolf.games as any).getSupportedFeatures = getSupportedFeatures;
  (overwolf.games as any).getSupportedFeaturesOnline = getSupportedFeaturesOnline;
}


index.js
'use strict';
if (typeof exports !== 'undefined') {
    Object.defineProperty(exports, '__esModule', { value: true });
}

/**
 * An enum that contains all games that support events through Overwolf
 */
const OverwolfGameWithEventSupport = {
    APEX: 21566,
    CallOfDutyWarzone: 21626,
    CSGO: 7764,
    DOTA2: 7314,
    DOTAUnderlords: 21586,
    EscapeFromTarkov: 21634,
    FinalFantasyXIV: 6350,
    Fortnite: 21216,
    Hearthstone: 9898,
    HeroesOfTheStorm: 10624,
    LeagueOfLegends: 5426,
    LegendsOfRuneterra: 21620,
    Minecraft: 8032,
    MTGArena: 21308,
    Overwatch: 10844,
    PUBG: 10906,
    PUBGLite: 21598,
    PathOfExile: 7212,
    RainbowSixSiege: 10826,
    RocketLeague: 10798,
    Splitgate: 21404,
    StarCraft2: 5855,
    TeamfightTactics: 5426,
    Valorant: 21640,
    Warframe: 8954,
    WorldOfTanks: 6365,
    WorldOfWarcraft: 765,
    WorldOfWarships: 10746
};

/**
 * Enum that contains the supported events per game, these will automatically register with setRequiredFeatures when the game is detected
 */
const OverwolfGameSupportedEvents = {
    APEX: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'me',
        'team',
        'kill',
        'damage',
        'death',
        'revive',
        'match_state',
        'match_info',
        'inventory',
        'location',
        'match_summary',
        'roster',
        'rank',
        'kill_feed'
    ],
    CallOfDutyWarzone: [
        'gep_internal',
        'match_info',
        'game_info',
        'kill',
        'death',
        'assist',
        'roster'
    ],
    CSGO: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'match_info',
        'kill',
        'death',
        'assist',
        'headshot',
        'round_start',
        'match_start',
        'match_info',
        'match_end',
        'team_round_win',
        'bomb_planted',
        'bomb_change',
        'reloading',
        'fired',
        'weapon_change',
        'weapon_acquired',
        'info',
        'roster',
        'player_activity_change',
        'team_set',
        'replay',
        'counters',
        'mvp',
        'scoreboard',
        'kill_feed'
    ],
    DOTA2: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'game_state_changed',
        'match_state_changed',
        'match_detected',
        'daytime_changed',
        'clock_time_changed',
        'ward_purchase_cooldown_changed',
        'match_ended',
        'kill',
        'assist',
        'death',
        'cs',
        'xpm',
        'gpm',
        'gold',
        'hero_leveled_up',
        'hero_respawned',
        'hero_buyback_info_changed',
        'hero_boughtback',
        'hero_health_mana_info',
        'hero_status_effect_changed',
        'hero_attributes_skilled',
        'hero_ability_skilled',
        'hero_ability_used',
        'hero_ability_cooldown_changed',
        'hero_ability_changed',
        'hero_item_cooldown_changed',
        'hero_item_changed',
        'hero_item_used',
        'hero_item_consumed',
        'hero_item_charged',
        'match_info',
        'roster',
        'party',
        'error',
        'hero_pool',
        'me',
        'game'
    ],
    DOTAUnderlords: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'match_info'
    ],
    EscapeFromTarkov: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'match_info',
        'game_info'
    ],
    FinalFantasyXIV: [
        'match_info'
    ],
    Fortnite: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'kill',
        'killed',
        'killer',
        'revived',
        'death',
        'match',
        'match_info',
        'rank',
        'me',
        'phase',
        'location',
        'team',
        'items',
        'counters'
    ],
    Hearthstone: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'scene_state',
        'collection',
        'decks',
        'match',
        'match_info'
    ],
    HeroesOfTheStorm: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'match_info',
        'me',
        'game_info',
        'roster',
        'death',
        'kill'
    ],
    LeagueOfLegends: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'live_client_data',
        'matchState',
        'match_info',
        'death',
        'respawn',
        'abilities',
        'kill',
        'assist',
        'gold',
        'minions',
        'summoner_info',
        'gameMode',
        'teams',
        'level',
        'announcer',
        'counters',
        'damage',
        'heal'
    ],
    LegendsOfRuneterra: [
        'game_client_data'
    ],
    Minecraft: [
        'gep_internal',
        'game_info',
        'match_info'
    ],
    MTGArena: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'game_info',
        'match_info'
    ],
    Overwatch: [
        'gep_internal',
        'game_info',
        'match_info',
        'kill',
        'death'
    ],
    PUBG: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'kill',
        'revived',
        'death',
        'killer',
        'match',
        'rank',
        'counters',
        'location',
        'me',
        'team',
        'phase',
        'map',
        'roster'
    ],
    PUBGLite: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'kill',
        'revived',
        'death',
        'killer',
        'match',
        'rank',
        'me',
        'phase',
        'map',
        'team_feed',
        'team'
    ],
    PathOfExile: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'kill',
        'death',
        'me',
        'match_info'
    ],
    RainbowSixSiege: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'game_info',
        'match',
        'match_info',
        'roster',
        'kill',
        'death',
        'me'
    ],
    RocketLeague: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'stats',
        'teamGoal',
        'opposingTeamGoal',
        'match',
        'roster',
        'me',
        'match_info'
    ],
    Splitgate: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'game_info',
        'match_info',
        'player',
        'location',
        'match',
        'feed',
        'connection',
        'kill',
        'death',
        'portal',
        'assist'
    ],
    StarCraft2: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'match_info'
    ],
    TeamfightTactics: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'live_client_data',
        'me',
        'match_info',
        'roster',
        'store',
        'board',
        'bench',
        'carousel'
    ],
    Valorant: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'me',
        'game_info',
        'match_info',
        'kill',
        'death'
    ],
    Warframe: [
        'gep_internal',
        'game_info',
        'match_info'
    ],
    WorldOfTanks: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'kill',
        'death',
        'game_info',
        'match_info'
    ],
    WorldOfWarcraft: [
        'game_info'
    ],
    WorldOfWarships: [
        /**
         * Internal use for Overwolf (Version numbers for GEP, Game Event Provider)
         */
        'gep_internal',
        'game_info',
        'account_info',
        'match',
        'match_info',
        'kill',
        'death'
    ]
};

/**
 * @typedef {Object} OverwolfGameWithEvents
 * @property {string} game
 * @property {string[]} events
 */

/**
 * Gets the supported events for the specified classId
 * @param {Number} classId This is the classId you get from the gameInfo-object
 * @returns {OverwolfGameWithEvents} Returns an array of supported events if there's a game that corresponds to the classId, otherwise null
 */
function getSupportedFeatures (classId) {
    let game = Object.keys(OverwolfGameWithEventSupport).find(key => OverwolfGameWithEventSupport[key] == classId);
    if (game) {
        return { game: game, events: OverwolfGameSupportedEvents[game] };
    }

    return null;
}

 /**
     * Gets the supported events for the specified classId
     * @param {Number} classId This is the classId you get from the gameInfo-object
     * @param {Function|null} callback This is what you use to manage the data returned
     */
    function getSupportedFeaturesOnline(classId, callback) {
      let game = Object.keys(OverwolfGameWithEventSupport).find((key) => OverwolfGameWithEventSupport[key] == classId);
      if (game) {
          fetch(`https://game-events-status.overwolf.com/${classId}_prod.json`)
              .then((resp) => {
              return resp.json();
          })
              .then((json) => {
              if (!!callback) {
                  callback({
                      game: game,
                      events: json.features.map((f) => f.name),
                  });
              }
          })
              .catch((err) => console.log(err));
      }
  }

if (typeof overwolf !== "undefined" && typeof overwolf.games !== "undefined") {
    overwolf.games.getSupportedFeatures = getSupportedFeatures;
    overwolf.games.getSupportedFeaturesOnline = getSupportedFeaturesOnline;
}

if (typeof exports !== 'undefined') {
    exports.getSupportedFeatures = getSupportedFeatures;
    exports.getSupportedFeaturesOnline = getSupportedFeaturesOnline;

    exports.OverwolfGameWithEventSupport = OverwolfGameWithEventSupport;
    exports.OverwolfGameSupportedEvents = OverwolfGameSupportedEvents;
}



